src/cachemachine/automatedteller.py

```diff
@@ -0,0 +1,75 @@
+import asyncio
+from typing import Any, Dict, List, Sequence, Union
+
+import structlog
+
+from cachemachine.cachechecker import CacheChecker
+from cachemachine.cachedepositer import CacheDepositer
+from cachemachine.rubinrepoman import RubinRepoMan
+from cachemachine.simplerepoman import SimpleRepoMan
+
+logger = structlog.get_logger(__name__)
+
+
+class AutomatedTeller:
+    available_images: List[Dict[str, str]] = []
+    desired_images: List[Dict[str, str]] = []
+    images_to_cache: List[Dict[str, str]] = []
+
+    def __init__(
+        self,
+        name: str,
+        labels: Dict[str, str],
+        repomen: Sequence[Union[SimpleRepoMan, RubinRepoMan]],
+    ):
+        self.name = name
+        self.labels = labels
+        self.repomen = repomen
+
+        self.recommended_image_urls = []
+        for r in self.repomen:
+            if r.recommended_image_url:
+                self.recommended_image_urls.append(r.recommended_image_url)
+
+        self.checker = CacheChecker(self.labels, self.recommended_image_urls)
+        self.depositer = CacheDepositer(self.name, self.labels)
+
+    # Note, doesn't actually return, intended to run forever.
+    async def do_work(self) -> None:
+        while True:
+            self.available_images = []
+            self.desired_images = []
+            self.images_to_cache = []
+
+            self.checker.check()
+
+            if not self.checker.nodes_exist:
+                logger.warning(f"No nodes are labeled with: {self.labels}")
+            else:
+                for r in self.repomen:
+                    recommended_names = self.checker.recommended_names[
+                        r.recommended_image_url
+                    ]
+                    for image in r.desired_images(recommended_names):
+                        self.desired_images.append(image)
+
+                        if image["image_url"] in self.checker.common_cache:
+                            self.available_images.append(image)
+                        else:
+                            self.images_to_cache.append(image)
+
+            if self.images_to_cache and not self.depositer.busy():
+                self.depositer.deposit(self.images_to_cache[0]["image_url"])
```

Comment:
I don't understand why it would be pulling six images at once. Wouldn't you pull one image, wait for it to finish, and then pull the next image, so it would only be pulling one image at a time? In other words, I think I'm arguing for exactly what you said at the end of the above paragraph, but that's not what the code does.

Reply:
"Could you explain the rationale for only caching one image at a time and not caching all of the missing ones on each pass through the loop?"

I was answering your question why we shouldn't do that.  I do pull an image, wait, and pull the next one.  That's what I think the code does.  How does it not?

Reply:
Maybe I'm misunderstanding the code, but what I think it does is:

1. Calculate which images are missing from the relevant nodes.
2. Select the first of those images.
3. See if the depositor is busy. If it is, do nothing. If it's not, start caching that first image.
4. Wait for a minute.
5. Discard the list of images and start again.

What I was expecting was:

1. Calculate which images are missing from the relevant nodes.
2. For each image, tell the depositor to cache it and wait for it to finish (probably by making deposit an async method that internally waits and checks until the deployment is done, which would remove the need for the busy method).
3. Wait until it's possible the list of desired images or the cache status of the nodes have changed and then repeat.

I think both algorithms get to the same end state, but I think the current code does a bit of extra work recalculating the list of images to cache every minute and I found it a bit harder to follow.

Reply:
Okay.  I thought for a minute you were saying there was a bug where the code doesn't do what we agree on it does here.  I agree that both of these ways end up at the same state.

There are two subtle things that are a little better in my slightly more complex approach:
1. If you restart or redeploy cachemachine, it finds the daemonset from before, and lets it finish which is really nice.
2. One of the whole reasons why we have this complicated docker scanning thing is because Adam likes to be able to push images and have them show up right away.  At least this would detect there's a new experimental or new recommended and pull that before any other work going on.  It is constantly pulling the most important thing, and higher priority images won't get starved out by lower priority ones.

I'll try to make this all more clear.